index.html

```diff
@@ -204,55 +204,1951 @@ <h3>
+        <h3>
+          <dfn>getDevices()</dfn> method
+        </h3>
+        <p>
+          The {{HID/getDevices()}} method steps are:
+        </p>
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          </li>
+          <li>If [=this=]'s [=relevant global object=]'s [=associated
+          Document=] is not [=allowed to use=] the [=policy-controlled
+          feature=] named `"hid"`, [=reject=] |promise| with a
+          "{{SecurityError}}" {{DOMException}} and return |promise|.
+          </li>
+          <li>Run the following steps [=in parallel=]:
+            <ol>
+              <li>Let |devices| be an empty [=list=].
+              </li>
+              <li>Repeat for each |device:HIDDevice| in
+              |navigator:Navigator|.{{Navigator/[[hidDevices]]}}:
+                <ol>
+                  <li>If the user has allowed the site to access |device| as
+                  the result of a previous call to {{HID/requestDevice()}},
+                  then append |device| to |devices|.
+                  </li>
+                </ol>
+              </li>
+              <li>[=Queue a global task=] on the [=relevant global object=] of
+              [=this=] using the [=HID device task source=] to [=resolve=]
+              |promise| with |devices|.
+              </li>
+            </ol>
+          </li>
+          <li>Return |promise|.
+          </li>
+        </ol>
+      </section>
+      <section>
+        <h3>
+          <dfn>requestDevice()</dfn> method
+        </h3>
+        <p>
+          The {{HID/requestDevice()}} method steps are:
+        </p>
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          </li>
+          <li>If [=this=]'s [=relevant global object=]'s [=associated
+          Document=] is not [=allowed to use=] the [=policy-controlled
+          feature=] named `"hid"`, [=reject=] |promise| with a
+          "{{SecurityError}}" {{DOMException}} and return |promise|.
+          </li>
+          <li>If the [=relevant global object=] of [=this=] does not have
+          [=transient activation=], [=reject=] |promise| with a
+          "{{SecurityError}}" {{DOMException}} and return |promise|.
+          </li>
+          <li>If |options|["{{HIDDeviceRequestOptions/filters}}"] is present,
+          then run the following steps for each |filter:HIDDeviceFilter| in
+          |options|["{{HIDDeviceRequestOptions/filters}}"]:
+            <ol>
+              <li>If |filter|["{{HIDDeviceFilter/productId}}"] is present and
+              |filter|["{{HIDDeviceFilter/productId}}"] is not present,
+              [=reject=] |promise| with a {{TypeError}} and return |promise|.
+              </li>
+              <li>If |filter|["{{HIDDeviceFilter/usage}}"] is present and
+              |filter|["{{HIDDeviceFilter/usagePage}}"] is not present,
+              [=reject=] |promise| with a {{TypeError}} and return |promise|.
+              </li>
+            </ol>
```

Comment:
Is an empty filter (not an empty list of filters) valid?

Reply:
Yes, an empty filter should match any device

Reply:
What do you think about making an empty filter throw a TypeError? A developer can match any device by providing an empty list of filters.

Reply:
Sounds good, now we'll reject with TypeError. I'll leave this open as a reminder to file a spec bug against Chrome.